### PR TITLE
rgw multisite: fixes for concurrent version creation

### PR DESCRIPTION
**Pull request by cbodley**
this patch set makes adjustments to object versioning so that all peer zones converge on the same current version, even if those zones are creating versions concurrently and replicating them in a different order

the existing algorithm uses an olh epoch, incremented with each new version of a name, that is used to sort its versions from newest to oldest. and the last version written becomes the 'current' version

rather than always promoting the last version written to 'current', we have to detect olh epoch collisions, and use the same sort order as the bucket index to decide whether a version really is the latest one. this change is made both in cls_rgw when setting/clearing the RGW_BUCKET_DIRENT_FLAG_CURRENT flag on bucket index entries, and in radosgw when processing the olh log to update the object logical head

Fixes: https://tracker.ceph.com/issues/39142

### CODE REVIEW

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -6781,28 +6781,59 @@ int RGWRados::apply_olh_log(RGWObjectCtx& obj_ctx, RGWObjState& state, const RGW
...
   bool delete_marker = false;
   list<cls_rgw_obj_key> remove_instances;
   bool need_to_remove = false;
 
+  // decode current epoch and instance
+  auto olh_ver = state.attrset.find(RGW_ATTR_OLH_VER);
+  if (olh_ver != state.attrset.end()) {
+    std::string str = olh_ver->second.to_str();
+    std::string err;
+    link_epoch = strict_strtoll(str.c_str(), 10, &err);
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: need to check `!err.empty()`?

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -6781,28 +6781,59 @@ int RGWRados::apply_olh_log(RGWObjectCtx& obj_ctx, RGWObjState& state, const RGW
...
+  }
+  auto olh_info = state.attrset.find(RGW_ATTR_OLH_INFO);
+  if (olh_info != state.attrset.end()) {
+    RGWOLHInfo info;
+    try {
+      auto p = olh_info->second.cbegin();
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: This is a copy-paste of code off `get_olh()`, can move it to a common method

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
@smanjara this is the PR i mentioned during our discussion of https://tracker.ceph.com/issues/58939

**Comment by smanjara** [comment]
> @smanjara this is the PR i mentioned during our discussion of https://tracker.ceph.com/issues/58939

thank you!

**Pull request merged**